analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs

```diff
@@ -153,34 +153,38 @@ public override void ExecutionCompleted()
 
     private ProgramState ProcessInvocation(SymbolicContext context, IInvocationOperationWrapper invocation)
     {
-        return invocation.TargetMethod.Is(KnownType.System_Linq_Enumerable, nameof(Enumerable.Count))
-            && SizeConstraint(context.State, invocation.Instance ?? invocation.Arguments[0].ToArgument().Value, HasFilteringPredicate()) is { } constraint
-                ? context.SetOperationConstraint(constraint)
-                : ProcessMethod(context, invocation.TargetMethod, invocation.Instance);
-
-        bool HasFilteringPredicate() =>
-            invocation.Arguments.Any(x => x.ToArgument().Parameter.Type.Is(KnownType.System_Func_T_TResult));
-    }
-
-    private ProgramState ProcessMethod(SymbolicContext context, IMethodSymbol method, IOperation instance)
-    {
-        var state = context.State;
-        if (instance is null)
+        if (invocation.TargetMethod.Is(KnownType.System_Linq_Enumerable, nameof(Enumerable.Count))
+            && SizeConstraint(context.State, invocation.Instance ?? invocation.Arguments[0].ToArgument().Value, HasFilteringPredicate()) is { } constraint)
         {
-            return state;
+            return context.SetOperationConstraint(constraint);
         }
-        if (RaisingMethods.Contains(method.Name))
+        else if (invocation.Instance is { } instance)
         {
-            if (state[instance]?.HasConstraint(CollectionConstraint.Empty) is true)
-            {
-                emptyAccess.Add(context.Operation.Instance);
-            }
-            else
+            if (RaisingMethods.Contains(invocation.TargetMethod.Name))
             {
-                nonEmptyAccess.Add(context.Operation.Instance);
+                if (context.State[instance]?.HasConstraint(CollectionConstraint.Empty) is true)
+                {
+                    emptyAccess.Add(context.Operation.Instance);
+                }
+                else
+                {
+                    nonEmptyAccess.Add(context.Operation.Instance);
+                }
             }
+            return ProcessAddMethod(context.State, invocation.TargetMethod, instance);
+        }
+        else
+        {
+            return context.State;
         }
-        if (AddMethods.Contains(method.Name))
+
+        bool HasFilteringPredicate() =>
+            invocation.Arguments.Any(x => x.ToArgument().Parameter.Type.Is(KnownType.System_Func_T_TResult));
+    }
+
+    private static ProgramState ProcessAddMethod(ProgramState state, IMethodSymbol method, IOperation instance)
+    {
+        if (instance is not null && AddMethods.Contains(method.Name))
```

Comment:
Move the `instance is not null` check to the call-side for method references. `ProcessInvocation` has the check already.